csrc/scheduler/mma_utils.cpp

```diff
@@ -1051,154 +1054,103 @@ inline void resolveTvToMatmulDomainsMapping(
-      if (km_found) {
+    std::optional<MatmulDomain> group_inner_dom = std::nullopt;
+    for (TensorView* tv : role_it->second) {
+      IterDomain* inner_id =
+          TensorDomain::noReductions(tv->getMaybeAllocationDomain()).back();
+      const ValGroup& g = exact_graph.toGroup(inner_id);
+      auto g_it = dim_roles.find(g);
+      if (g_it == dim_roles.end()) {
         return {
-            "Failed to find MMA input, more than one fusion input has [..., K, ..., M, ...] iter domains"};
+            MatmulDomain::M,
+            "Inner domain of tensor was not mapped to a MatmulDomain"};
       }
-      km_found = true;
-    }
-    if (item.second == nk_desc) {
-      if (nk_found) {
+      if (!group_inner_dom.has_value()) {
+        group_inner_dom = g_it->second;
+      } else if (group_inner_dom.value() != g_it->second) {
         return {
-            "Failed to find MMA input, more than one fusion input has [..., N, ..., K, ...] iter domains"};
+            MatmulDomain::M, "Group contains multiple inner dimension domains"};
       }
-      nk_found = true;
     }
-    if (item.second == kn_desc) {
-      if (kn_found) {
-        return {
-            "Failed to find MMA input, more than one fusion input has [..., K, ..., N, ...] iter domains"};
-      }
-      kn_found = true;
+    if (!group_inner_dom.has_value()) {
+      return {MatmulDomain::M, "No tensor found in role"};
     }
-  }
+    return {group_inner_dom.value(), ""};
+  };
 
-  if ((mk_found && kn_found) && !(km_found || nk_found)) {
-    return MmaLayout::TT;
+  const InnerDomResult a_dom_res = innerDomain(MatmulRole::INPUT_A);
+  if (!a_dom_res.second.empty()) {
+    std::string err = a_dom_res.second;
+    return err;
   }
-  if ((km_found && kn_found) && !(mk_found || nk_found)) {
-    return MmaLayout::NT;
+  const bool kinner_a = a_dom_res.first == MatmulDomain::K;
+
+  const InnerDomResult b_dom_res = innerDomain(MatmulRole::INPUT_B);
+  if (!b_dom_res.second.empty()) {
+    std::string err = b_dom_res.second;
+    return err;
   }
-  if ((mk_found && nk_found) && !(km_found || kn_found)) {
+  const bool kinner_b = b_dom_res.first == MatmulDomain::K;
+
+  if (kinner_a && kinner_b) {
     return MmaLayout::TN;
-  }
-  if ((km_found && nk_found) && !(mk_found || kn_found)) {
+  } else if (kinner_a && !kinner_b) {
+    return MmaLayout::TT;
+  } else if (!kinner_a && !kinner_b) {
+    return MmaLayout::NT;
+  } else if (!kinner_a && kinner_b) {
     return MmaLayout::NN;
   }
-
-  return {"Failed to decide fusion inputs' data layout."};
+  NVF_ERROR(false, "Reached unreachable section of getProblemLayout");
 }
 
-MatmulProblemLayoutOpt getMmaLayout(Fusion* fusion) {
-  auto mma_exprs = ir_utils::getOpsOfType<MmaOp>(fusion);
-  if (mma_exprs.size() != 1) {
-    std::stringstream ss;
-    ss << "Invalid number of MmaOp instances in fusion, expected 1, got "
-       << mma_exprs.size();
-    return ss.str();
+MatmulProblemLayoutOpt getProblemLayout(
```

Comment:
Good catch! I refactored it as you suggest and removed the `getProblemLayout(Fusion*, const MatmulPattern&)` signature. I don't love the `getProblemLayout(Fusion*)` signature either since we should encourage reusing data when possible instead of rebuilding the entire `IdModel`. I think that signature is only used in tests but it seems to be useful for now so I kept it.